### DISCUSSION TITLE
Prime Blocks: styles not defined

### DISCUSSION THREAD

**Discussion by ahallez** [primefaces/community#2207] [PrimeVue Blocks]
Hello,

I'm trying out primevue with prime blocks for the next version of my application, but I'm having some trouble figuring out how to get all of the css included in my test page.

Here is what I've done so far:

installed primevue 4 and related components:

[package.json](https://github.com/user-attachments/files/15985240/package.json)

Set up my main.ts file to use primevue and included primeflex css and primeicons:

```
import { createApp } from 'vue'
import { createPinia } from 'pinia'
import PrimeVue from 'primevue/config'
import Aura from '@primevue/themes/aura'
import Ripple from 'primevue/ripple'
import StyleClass from 'primevue/styleclass'

import "primeflex/primeflex.css"
import "primeicons/primeicons.css"

import App from './App.vue'
import router from './router'

const app = createApp(App)

app.use(createPinia())
app.use(router)
app.use(PrimeVue, {
  theme: {
    preset: Aura,
    options: {
      prefix: 'p',
      cssLayer: true,
    },
    ripple: true,
  },
})
app.directive('ripple', Ripple)
app.directive('styleclass', StyleClass)

app.mount('#app')

```
Then pasted a block into my App.vue file (the stacked layout with Colored topbar)

```
<script setup lang="ts">
</script>

<template>
  <div class="min-h-screen flex flex-column">
    <div class="bg-primary py-3 px-5 flex align-items-center justify-content-between relative lg:static" style="min-height: 80px">
      <img src="/src/assets/logo.svg" alt="Image" height="40" class="mr-0 lg:mr-6" />
      <a v-ripple class="cursor-pointer block lg:hidden text-indigo-50 p-ripple"
         v-styleclass="{ selector: '@next', enterClass: 'hidden', leaveToClass: 'hidden', hideOnOutsideClick: true }">
        <i class="pi pi-bars text-4xl"></i>
      </a>
      <div class="align-items-center flex-grow-1 justify-content-between hidden lg:flex absolute lg:static w-full bg-indigo-500 left-0 top-100 z-1 shadow-2 lg:shadow-none">
        <ul class="list-none p-0 m-0 flex lg:align-items-center select-none flex-column lg:flex-row">
          <li>
            <a v-ripple class="flex px-6 p-3 lg:px-3 lg:py-2 align-items-center text-indigo-100 hover:text-indigo-50 hover:bg-indigo-600 font-medium border-round cursor-pointer transition-colors transition-duration-150 p-ripple">
              <i class="pi pi-home mr-2"></i>
              <span>Home</span>
            </a>
          </li>
          <li>
            <a v-ripple class="flex px-6 p-3 lg:px-3 lg:py-2 align-items-center text-indigo-100 hover:text-indigo-50 hover:bg-indigo-600 font-medium border-round cursor-pointer transition-colors transition-duration-150 p-ripple"
               v-styleclass="{ selector: '@next', enterClass: 'hidden', enterActiveClass: 'scalein', leaveToClass: 'hidden', leaveActiveClass: 'fadeout', hideOnOutsideClick: true }">
              <i class="pi pi-users mr-2"></i>
              <span>Customers</span>
              <i class="pi pi-angle-down ml-auto lg:ml-3"></i>
            </a>
            <ul class="list-none py-3 px-6 m-0 lg:px-0 lg:py-0 border-round shadow-0 lg:border-1 border-50 lg:shadow-2 lg:absolute surface-overlay hidden origin-top w-full lg:w-15rem cursor-pointer">
              <li>
                <a v-ripple class="flex p-3 align-items-center text-600 hover:text-900 hover:surface-100 transition-colors transition-duration-150 p-ripple">
                  <i class="pi pi-user-plus mr-2"></i>
                  <span class="font-medium">Add New</span>
                </a>
              </li>
              <li class="relative">
                <a v-ripple class="flex p-3 align-items-center text-600 hover:text-900 hover:surface-100 transition-colors transition-duration-150 p-ripple"
                   v-styleclass="{ selector: '@next', enterClass: 'hidden', enterActiveClass: 'scalein', leaveToClass: 'hidden', leaveActiveClass: 'fadeout', hideOnOutsideClick: true }">
                  <i class="pi pi-search mr-2"></i>
                  <span class="font-medium">Search</span>
                  <i class="pi pi-angle-down ml-auto lg:-rotate-90"></i>
                </a>
                <ul class="list-none py-3 pl-3 m-0 lg:px-0 lg:py-0 border-round shadow-0 lg:border-1 border-50 lg:shadow-2 lg:absolute surface-overlay hidden origin-top w-full lg:w-15rem cursor-pointer left-100 top-0">
                  <li>
                    <a v-ripple class="flex p-3 align-items-center text-600 hover:text-900 hover:surface-100 transition-colors transition-duration-150 p-ripple">
                      <i class="pi pi-shopping-cart mr-2"></i>
                      <span class="font-medium">Purchases</span>
                    </a>
                  </li>
                  <li class="relative">
                    <a v-ripple class="flex p-3 align-items-center text-600 hover:text-900 hover:surface-100 transition-colors transition-duration-150 p-ripple">
                      <i class="pi pi-comments mr-2"></i>
                      <span class="font-medium">Messages</span>
                    </a>
                  </li>
                </ul>
              </li>
            </ul>
          </li>
          <li>
            <a v-ripple class="flex px-6 p-3 lg:px-3 lg:py-2 align-items-center text-indigo-100 hover:text-indigo-50 hover:bg-indigo-600 font-medium border-round cursor-pointer transition-colors transition-duration-150 p-ripple">
              <i class="pi pi-calendar mr-2"></i>
              <span>Calendar</span>
            </a>
          </li>
          <li>
            <a v-ripple class="flex px-6 p-3 lg:px-3 lg:py-2 align-items-center text-indigo-100 hover:text-indigo-50 hover:bg-indigo-600 font-medium border-round cursor-pointer transition-colors transition-duration-150 p-ripple">
              <i class="pi pi-chart-line mr-2"></i>
              <span>Stats</span>
            </a>
          </li>
        </ul>
        <ul class="list-none p-0 m-0 flex lg:align-items-center select-none flex-column lg:flex-row border-top-1 border-indigo-400 lg:border-top-none">
          <li>
            <a v-ripple class="flex px-6 p-3 lg:px-3 lg:py-2 align-items-center text-indigo-100 hover:text-indigo-50 hover:bg-indigo-600 font-medium border-round cursor-pointer transition-colors transition-duration-150 p-ripple">
              <i class="pi pi-inbox text-base lg:text-2xl mr-2 lg:mr-0"></i>
              <span class="block lg:hidden font-medium">Inbox</span>
            </a>
          </li>
          <li>
            <a v-ripple class="flex px-6 p-3 lg:px-3 lg:py-2 align-items-center text-indigo-100 hover:text-indigo-50 hover:bg-indigo-600 font-medium border-round cursor-pointer transition-colors transition-duration-150 p-ripple">
              <i class="pi pi-bell text-base lg:text-2xl mr-2 lg:mr-0"></i>
              <span class="block lg:hidden font-medium">Notifications</span>
            </a>
          </li>
          <!--<li class="border-top-1 border-indigo-400 lg:border-top-none">
            <a v-ripple class="flex px-6 p-3 lg:px-3 lg:py-2 align-items-center hover:bg-indigo-600 font-medium border-round cursor-pointer transition-colors transition-duration-150 p-ripple">
              <img src="/images/blocks/avatars/circle/avatar-f-1.png" class="mr-3 lg:mr-0" style="width: 32px; height: 32px"/>
              <div class="block lg:hidden">
                <div class="text-indigo-50 font-medium">Josephine Lillard</div>
                <span class="text-indigo-100 font-medium text-sm">Marketing Specialist</span>
              </div>
            </a>
          </li>-->
        </ul>
      </div>
    </div>
    <ul class="list-none py-3 px-5 m-0 surface-section flex align-items-center font-medium overflow-x-auto border-top-1 border-bottom-1 surface-border shadow-2 lg:shadow-none">
      <li class="pr-3">
        <a class="cursor-pointer">
          <i class="pi pi-home text-blue-500"></i>
        </a>
      </li>
      <li class="px-2">
        <i class="pi pi-angle-right text-500"></i>
      </li>
      <li class="px-2">
        <a class="cursor-pointer text-blue-500 white-space-nowrap">Level 3</a>
      </li>
      <li class="px-2">
        <i class="pi pi-angle-right text-500"></i>
      </li>
      <li class="px-2">
        <a class="cursor-pointer text-blue-500 white-space-nowrap">Level 2</a>
      </li>
      <li class="px-2">
        <i class="pi pi-angle-right text-500"></i>
      </li>
      <li class="px-2">
        <a class="cursor-pointer text-900 white-space-nowrap">Level 1</a>
      </li>
    </ul>
    <div class="p-5 flex flex-column flex-auto">
      <div class="border-2 border-dashed surface-border border-round surface-section flex-auto">
        <RouterView />
      </div>
    </div>
  </div>
</template>

<style lang='scss'>
</style>

```

This is what the page looks like:

Most notably variables such as `var(--indigo-500)` are not defined

![image](https://github.com/primefaces/community/assets/12940726/619e6803-e833-44a6-b74d-599207a4033d)

**Comment by vieira-brz** [comment]
It looks like you are trying to integrate PrimeVue components with styling from PrimeFlex and possibly a custom theme, but you're facing issues with some CSS variables not being defined. This is a common issue when styles are not fully integrated or when the theme does not define all the necessary variables.

Here’s how you can troubleshoot and potentially fix this issue:

### 1. **Ensure Theme Integration**

First, make sure that the theme you are using (Aura in your case) is correctly integrated and that it includes definitions for all the colors and other style properties used in your component. If `var(--indigo-500)` and other similar variables are not defined, it might indicate that the theme files are not properly loaded or are missing some definitions.

- **Check the Theme File**: Locate the theme CSS file in your project (usually under `node_modules/@primevue/themes`) and check if the color variables like `--indigo-500` are defined there. If they are not, you might need to define these yourself.

### 2. **Add Missing CSS Variables**

If the theme does not include all the variables you need, you can add them manually in your global CSS or in the component's scoped style:

```css
/* In your global styles or near the top of your component styles */
:root {
    --indigo-500: #6666ff;  /* Adjust the color code as needed */
    --indigo-100: #ededff;  /* Example for lighter indigo */
    --indigo-600: #5656e6;  /* Example for darker indigo */
    /* Add other missing variables similarly */
}
```

### 3. **Importing CSS Files**

Verify that all necessary CSS files are correctly imported in your project setup. Since you are using Vue and Vite/Webpack, ensure that the imports in your `main.ts` or elsewhere are correctly resolving:

```javascript
import "primevue/resources/themes/aura/theme.css"; // Adjust path if necessary
import "primeflex/primeflex.css";
import "primeicons/primeicons.css";
```

- **Note**: The path to the theme CSS might vary depending on how your project is structured or where your node_modules is located. Ensure the path accurately points to the correct file.

### 4. **Inspect Your Build Configuration**

Ensure that your build tool (like Vite or Webpack) is correctly configured to include CSS files from `node_modules`. Sometimes, build configurations can exclude these directories or mishandle asset resolutions.

### 5. **Check Console for Errors**

When you load your page, check the browser's developer console for any errors related to loading CSS files or other resources. Network tab errors or 404s can give you clues about what might not be loading properly.

### 6. **Further Debugging**

- **Development Tools**: Use browser developer tools to inspect elements that are missing styles. Check if the CSS variables are being inherited from `:root` or if there are other overriding styles.
- **Clear Cache**: Sometimes, browsers might cache old CSS files. Clear your browser cache or try in incognito mode to ensure you are loading the latest styles.

By following these steps, you should be able to resolve issues related to missing CSS variables and ensure your application styles are applied correctly. If the issue persists, consider checking if there are any updates or patches to the theme or PrimeVue components that might address these issues.

**Reply by ahallez** [reply]
Hi @vieira-brz,

Thanks for your reply. However it looks to me like you are referring to primevue v3 style of including css.

With the v4 theme, I can't seem to find any particular css file that could be included. When I check the example projects (https://github.com/primefaces/primevue-examples), they all use the Aura theme and none of those import theme related css.

From studying the examples, it was my impression that loading the theme with primevue somehow created the theme related css variables.

Maybe the conclusion is that the Primevue Blocks of Primeflex code is not updated to be used with primevue v4 then?

**Reply by vieira-brz** [reply]
Hmm, could be, but try this first:

In PrimeVue v4, themes are integrated differently. Ensure you have the correct theme package installed and imported in your project. For example, to use the Aura theme, you would do:

   ```javascript
   // main.ts or main.js
   import { createApp } from 'vue';
   import App from './App.vue';
   import PrimeVue from 'primevue/config';

   // Importing the theme
   import "primevue/resources/themes/aura/theme.css";  // Make sure the path is correct
   import "primevue/resources/primevue.min.css";
   import "primeicons/primeicons.css";
   import "primeflex/primeflex.css";

   const app = createApp(App);
   app.use(PrimeVue);
   app.mount('#app');
   ```
   
If certain variables like `--indigo-500` are still missing, you might need to define them manually. You can add them in your global CSS file or as a part of your component styles:

   ```css
   /* In your global styles (e.g., styles.css) */
   :root {
       --indigo-500: #6666ff;  /* Adjust the color code as needed */
       --indigo-100: #ededff;  /* Example for lighter indigo */
       --indigo-600: #5656e6;  /* Example for darker indigo */
       /* Add other missing variables similarly */
   }
   ```
   
 Ensure that you are using compatible versions of PrimeVue and PrimeFlex. Incompatibilities can sometimes lead to missing styles.

   ```json
   // package.json
   {
     "dependencies": {
       "primevue": "^4.0.0",
       "primeflex": "^3.0.0",
       // other dependencies
     }
   }
   ```

**Reply by ahallez** [reply]
I found the issue to be with the prefix configured in the theme options. The primeflex classes refer to variables without prefix.
So you have to set the theme prefix to '' in order to get compatible variable names

**Reply by yury-hos** [reply]
I am having a similar issue. Setting prefix to '' did the trick for me for some variables (like colors, e.g. --cyan-600). 
But some variables are still missing, --surface-border for example. 
I'm using Aura theme with primevue 4. Are variables like --surface-border and --surface-ground even defined in primevue theme?

**Comment by almirhodzic** [comment]
I have the same issue as “viera-brz” with the style blocks and the use of PrimeVue 4. When I use the workaround from “ahallez”, I get close to the appearance of design blocks as seen in the online examples of PrimeVue. However, I then encounter problems with v-ripple and styleClass in the blocks that I copy and paste. The v4 components (without design blocks) work excellently and without any issues. I’m concerned that the design blocks are not compatible with PrimeVue 4, or am I doing something completely wrong?

Let me know if you need any further adjustments!
Here is the screenshot (note the v-ripple styleClass issues in console)
<img width="1338" alt="Bildschirmfoto 2024-06-29 um 07 02 56" src="https://github.com/primefaces/community/assets/24458972/d0041ecc-aa55-434e-96a8-e3650446e39b">

**Reply by ahallez** [reply]
For the ripple and styleclass to work (and badge as well), you need to include the directives (see the main.ts code in my original message above).
This hint was also helpful to get the styleclass transitions to work:
[StyleClass: deprecate enterClass/leaveClass use enterFromClass/leaveFromClass](https://github.com/primefaces/primevue/issues/4866)
